### PR TITLE
Zuora review

### PR DESCRIPTION
**Pull request by dgrinenko**
BatchSource implementation:
- [x] Implement core
- [x] Implement Object parsing
- [x] Implement Transformers
- [x] Implement Splits
- [x] Add objects
- [x] Validators
- [ ] Design doc update
- [ ] Testing & Fixing small bugs
- [ ] Integration tests

### CODE REVIEW

File: docs/Zuora-batchsource.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# SendGrid batch source
+
+Description
+-----------
+Plugin fetches data from SendGrid. SendGrid is a cloud-based service that assists businesses with email delivery. For 
+the end user SendGrid provides information about existing Marketing Campaigns, Email Analytics, Bounces, Spam Reports.
+ 
+Properties
+----------
+### General
```
Comment (aonischuk) [inline docs/Zuora-batchsource.md]: Every property description should end with a dot.

File: docs/Zuora-batchsource.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+**Reference Name:** Name used to uniquely identify this source for lineage, annotating metadata, etc.
+
+**Authentication type:** The way, how user would like to be authenticated to the SendGrid account
```
Comment (aonischuk) [inline docs/Zuora-batchsource.md]: add a brief description of types.

File: docs/Zuora-batchsource.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+- Stats
+  - Global Stats
+  - Category Stats
+  - Advanced Stats
+- Suppressions
+  - Bounces
+  - Global Unsubscribes
+  - Group Unsubscribes
+  
+**Data Source Fields:** The list of fields available for the retrieval
```
Comment (aonischuk) [inline docs/Zuora-batchsource.md]: Property descriptions should not start with "the", "a", "an", etc

File: docs/Zuora-batchsource.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+- *SingleSends:* id, name, status, created_at, updated_at,
+- *Senders:* id, nickname, address, address_2, city, country, state, zip, locked, created_at, updated_at, from(email, name), verified(status, reason), reply_to(name, email)
+- *Contacts:* id, first_name, last_name, list_ids, created_at, updated_at, email, Segments, id, name, parent_list_id, created_at, updated_at, sample_updated_at, contacts_count
+- *GlobalStats:* date ,blocks ,bounce_drops ,bounces ,clicks ,deferred ,invalid_emails ,opens ,processed ,requests ,spam_report_drops ,spam_reports ,unique_clicks ,unique_opens ,unsubscribe_drops ,unsubscribes
+- *CategoryStats:* name, type, date, blocks, bounce_drops, bounces, clicks, deferred, delivered, invalid_emails, opens, processed, requests, spam_report_drops, spam_reports, unique_clicks, unique_opens, unsubscribe_drops, unsubscribes
+- *AdvancedStats:* name, type, date, clicks, opens, unique_clicks, unique_opens
+- *Bounces:* created, email, reason, status
+- *GroupUnsubscribes:* id ,name ,description ,is_default ,last_email_send_at ,unsubscribes
+
+**Start Date:** The date in format YYYY-MM-DD, starting from which the data is requested
```
Comment (aonischuk) [inline docs/Zuora-batchsource.md]: same

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        <dependency>
+            <groupId>commons-logging</groupId>
+            <artifactId>commons-logging</artifactId>
+            <version>${common.logging.version}</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpcore</artifactId>
+            <version>4.4.12</version>
```
Comment (aonischuk) [inline pom.xml]: let's move all versions to properties.

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.cdap.plugin</groupId>
+    <artifactId>zuora</artifactId>
+    <version>1.3.0-SNAPSHOT</version>
+    <packaging>jar</packaging>
+    <description>Zuora Plugin</description>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
```
Comment (aonischuk) [inline pom.xml]: sort props in alphabetic order.

File: src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+    this.basicAuth = basicAuth;
+    this.client = HttpClients.custom()
+      .setUserAgent(CLIENT_NAME)
+      .build();
+
+    try {
+      JAXBContext jaxbContext = JAXBContext.newInstance(ZuoraDefinitions.class, ZuoraObjectSchema.class);
+      this.jaxbUnmarshaller = jaxbContext.createUnmarshaller();
+    } catch (JAXBException e) {
+      throw new RuntimeException(e);
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java]: add message.

File: src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+    HttpPost post = new HttpPost(String.format("%s/%s", apiEnpoint, OAUTH_URL));
+    List<NameValuePair> params = new ArrayList<>();
+    params.add(new BasicNameValuePair("client_id", this.clientId));
+    params.add(new BasicNameValuePair("client_secret", this.clientSecret));
+    params.add(new BasicNameValuePair("grant_type", OAUTH_GRANT_TYPE));
+
+    try {
+      post.setEntity(new UrlEncodedFormEntity(params));
+    } catch (UnsupportedEncodingException e) {
+      throw new IllegalArgumentException("client_id or client_secret have been passedd in ");
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java]: client_id or client_secret have been passedd in 
->
client_id or client_secret not have been passed in
->
client_id or client_secret were not provided

File: src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+      }
+    }
+  }
+
+  public ZuoraDefinitions getObjectList() throws IOException {
+    String data = makeApiRequest(Method.GET, REST_API_VERSION + "/" + DESCRIBE_API, null, null);
+    try (StringReader reader = new StringReader(data)) {
+      return (ZuoraDefinitions) jaxbUnmarshaller.unmarshal(reader);
+    } catch (JAXBException e) {
+      throw new IOException(e);
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java]: message

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/sink/ZuorSinkConfigValidator.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+ *  License for the specific language governing permissions and limitations under
+ *  the License.
+ */
+
+package io.cdap.plugin.zuora.plugin.batch.sink;
+
+/**
+ * Sink Config Validator
+ */
+public class ZuorSinkConfigValidator {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/sink/ZuorSinkConfigValidator.java]: I suppose you accidentally added empty files for sink in this PR. Let's remove them.

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/sink/ZuoraOutputFormat.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+ *  License for the specific language governing permissions and limitations under
+ *  the License.
+ */
+
+package io.cdap.plugin.zuora.plugin.batch.sink;
+
+/**
+ * Output Format
+ */
+public class ZuoraOutputFormat {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/sink/ZuoraOutputFormat.java]: same for all others

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraInputFormat.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+  @Override
+  public List<InputSplit> getSplits(JobContext jobContext) {
+    Configuration configuration = jobContext.getConfiguration();
+    ZuoraSourceConfig conf = gson.fromJson(
+      configuration.get(ZuoraSourceConfig.PROPERTY_CONFIG_JSON),
+      ZuoraSourceConfig.class
+    );
+
+    return conf.getObjectsWithSchema()
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraInputFormat.java]: please add detailed comment on how this is parallelized.
I assume for every object there is a new split?

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  @Override
+  public BaseObject getCurrentValue() {
+    return currentRecord;
+  }
+
+  @Override
+  public float getProgress() {
+    return 0.0f;
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java]: do we know the count? If so we should track progress here.

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordMultiReader.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+import io.cdap.plugin.zuora.restobjects.ObjectHelper;
+import io.cdap.plugin.zuora.restobjects.objects.BaseObject;
+import io.cdap.plugin.zuora.restobjects.objects.Pair;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Zuora Record Multi Reader
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordMultiReader.java]: please make this and other similar javadoc's a bit more broad.

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  @Override
+  public void close() {
+    //no-op
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java]: we need to close CloseableHttpClient ZuoraRestClient#client here?

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSource.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        .collect(Collectors.toList()));
+
+    batchSourceContext.setInput(Input.of(config.referenceName, new ZuoraInputFormatProvider(config)));
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+
+    IdUtils.validateReferenceName(config.referenceName, failureCollector);
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSource.java]: please move this into validate method itself.

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSourceConfig.java [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+  @Nullable
+  private String advObjectsToPull;
+
+  public ZuoraSourceConfig(String referenceName) {
+    super(referenceName);
+  }
+
+  @Override
+  protected void validate(FailureCollector failureCollector) {
+
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSourceConfig.java]: Why it this empty?
please do all possible validations. Like it authentication/connection/correctness of object names(if possible to verify), etc.

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSourceSplit.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/**
+ * Basic split on per-object basis
+ */
+public class ZuoraSourceSplit extends InputSplit implements Writable {
+  private Pair<String, String> objectKey;
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSourceSplit.java]: Can we create a more meaningful class with two fields instead of using a pair? Would be more readable.
Which would exactly represent what first and second field means.

File: src/main/java/io/cdap/plugin/zuora/restobjects/objects/BaseResult.java [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+  public String getNextPage() {
+    if (Strings.isNullOrEmpty(nextPage)) {
+      return null;
+    }
+
+    URL url;
+    try {
+      url = new URL("http://" + nextPage);
+    } catch (MalformedURLException e) {
+      return null;
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/restobjects/objects/BaseResult.java]: can you please explain this? So if cannot construct URL why do we return null?

File: src/main/java/io/cdap/plugin/zuora/restobjects/ObjectHelper.java [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+   * @param internalObjectName the name, provided via {@link ObjectDefinition#Name()}
+   * @param requestedFields    the names, provided via {@link SerializedName#value()}
+   * @return CDAP Schema
+   */
+  public static Schema buildSchema(List<String> internalObjectName, @Nullable List<String> requestedFields) {
+    return buildSchema(internalObjectName, requestedFields, false);
+  }
+
+  /**
+   * Provides schema definition for desired entities with only desired fields included
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/restobjects/ObjectHelper.java]: nice javadocs.

File: widgets/Zuora-batchsink.json [hunk 21]
```diff
@@ -0,0 +1,118 @@
+{
```
Comment (aonischuk) [inline widgets/Zuora-batchsink.json]: Would be better to review If source and sink would be different PRs I think.

File: src/main/java/io/cdap/plugin/zuora/plugin/common/BaseConfig.java [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+  @Macro
+  @Name(PROPERTY_AUTH_TYPE)
+  @Description("The way, how user would like to be authenticated to the account")
+  private String authType;
+
+  @Macro
+  @Name(PROPERTY_USERNAME_NAME)
+  @Description("Login name for the account")
+  @Nullable
+  private String authUserName;
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/plugin/common/BaseConfig.java]: UserName -> Username

File: widgets/Zuora-batchsource.json [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+      "widget-type": "non-editable-schema-editor",
+      "schema": {
+      }
+    }
+  ],
+  "filters": [
+    {
+      "name": "AuthTypeBasicFilter",
+      "condition": {
+        "expression": "authType == 'basic'"
```
Comment (aonischuk) [inline widgets/Zuora-batchsource.json]: According to Albert we should use named operators everywhere it's possible.
https://github.com/data-integrations/http/blob/develop/widgets/HTTP-batchsource.json#L523

File: src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java [hunk 24]
```diff
@@ -0,0 +1,380 @@
...
+        try {
+          if (Strings.isNullOrEmpty(x)) {
+            return;
+          }
+
+          arguments.keySet().stream()
+            .filter(x::equals)
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException(String.format(
+              "Object '%s' require %s argument, but nothing provided",
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java]: requires '%s'

File: src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java [hunk 25]
```diff
@@ -0,0 +1,380 @@
...
+   * @param arguments query arguments
+   * @throws IllegalArgumentException if any validation issue
+   */
+  private void checkIncomingArguments(ObjectInfo objectInfo, Map<String, String> arguments)
+    throws IllegalArgumentException {
+
+    if (objectInfo.getRequiredArguments() != null && !objectInfo.getRequiredArguments().isEmpty()) {
+      if (arguments == null || arguments.isEmpty()) {
+        throw new IllegalArgumentException(String.format(
+          "Object '%s' require input arguments to be passed, nothing found",
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java]: requires

File: src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java [hunk 26]
```diff
@@ -0,0 +1,380 @@
+/*
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/client/ZuoraRestClient.java]: Our goal is to make reviewing this as simple as possible for Google. Please as a description to PR here:
- add detailed general overview on how your code works in general. Like a high abstraction view.
- And also which classes/files are auto-generated to they can skip reviewing it 
- Also source and sink should be separate PR IMO, for simplicity of review.

File: pom.xml [hunk 27]
```diff
@@ -0,0 +1,362 @@
...
+        </dependency>
+        <dependency>
+            <groupId>antlr</groupId>
+            <artifactId>antlr</artifactId>
+            <version>2.7.7</version>
+            <scope>compile</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
```
Comment (aonischuk) [inline pom.xml]: please add a rat plugin.

File: src/main/java/io/cdap/plugin/zuora/RestObjectsConfig.java [hunk 28]
```diff
@@ -0,0 +1,1896 @@
+package io.cdap.plugin.zuora;
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zuora/RestObjectsConfig.java]: missing license header.

File: pom.xml [hunk 29]
```diff
@@ -0,0 +1,362 @@
...
+        <common.logging.version>1.2</common.logging.version>
+        <junit.version>4.11</junit.version>
+        <mockito.version>1.10.19</mockito.version>
+        <spark.version>2.1.3</spark.version>
+    </properties>
+
+    <scm>
+        <connection>scm:git:https://github.com/data-integrations/sendgrid.git</connection>
+        <developerConnection>scm:git@github.com:data-integrations/sendgrid.git</developerConnection>
+        <url>https://github.com/data-integrations/sendgrid</url>
```
Comment (aonischuk) [inline pom.xml]: will this be name sendgrid or zuora?

File: pom.xml [hunk 30]
```diff
@@ -0,0 +1,362 @@
...
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <cdap.version>6.1.0-SNAPSHOT</cdap.version>
+        <hadoop.version>2.8.0</hadoop.version>
+        <hydrator.version>2.3.0-SNAPSHOT</hydrator.version>
+        <gson.version>2.2.4</gson.version>
+        <common.logging.version>1.2</common.logging.version>
+        <junit.version>4.11</junit.version>
+        <mockito.version>1.10.19</mockito.version>
+        <spark.version>2.1.3</spark.version>
```
Comment (aonischuk) [inline pom.xml]: seems like this and some other are not used

File: pom.xml [hunk 31]
```diff
@@ -0,0 +1,362 @@
...
+                        <goals>
+                            <goal>check</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <dependencies>
+                    <dependency>
+                        <groupId>com.puppycrawl.tools</groupId>
+                        <artifactId>checkstyle</artifactId>
+                        <version>6.19</version>
```
Comment (aonischuk) [inline pom.xml]: https://github.com/data-integrations/jira-plugins/pull/2

**Review by aonischuk — CHANGES_REQUESTED** [review]

**Comment by aonischuk** [comment]
Also we should merge changes into develop. From feature/... branch.

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java (hunk 13 above)
Comment (dgrinenko) [reply src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java]: nope, there no information about amount of records in response, new "pages" processed while they comes

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java (hunk 15 above)
Comment (dgrinenko) [reply src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraRecordReader.java]: No, it is rest client - it is need to be initialized to cache the credentials, but it not keeping the connection up.

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSourceConfig.java (hunk 17 above)
Comment (dgrinenko) [reply src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraSourceConfig.java]: as in the plan mentioned, validation is not ready yet, working on it

File: pom.xml (hunk 30 above)
Comment (dgrinenko) [reply pom.xml]: they would be used with added integration tests

File: src/main/java/io/cdap/plugin/zuora/restobjects/objects/BaseResult.java (hunk 19 above)
Comment (dgrinenko) [reply src/main/java/io/cdap/plugin/zuora/restobjects/objects/BaseResult.java]: The default value for this property is null, which is returned by API. It means that no new page provided and server response is completed at last page. 

If the provided url by the server is malformed, it could be caused by an exception on back-end and should not ddos api with wrong api and non-existing page. (i.e. we considering wrong url as the last result page in the query)

Comment (dgrinenko) [reply src/main/java/io/cdap/plugin/zuora/restobjects/objects/BaseResult.java]: added comment to the code

File: src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraInputFormat.java (hunk 12 above)
Comment (dgrinenko) [reply src/main/java/io/cdap/plugin/zuora/plugin/batch/source/ZuoraInputFormat.java]: correct, added to the javadoc

**Comment by dgrinenko** [comment]
moved everything to `feature/zuora`
